### ISSUE TITLE
[Feature Request] Save States usability improvements

### ISSUE THREAD

**Issue by liandris** [libretro/RetroArch#13062]
1. Save States screen can be made more useful. Specifically, each save state can have a save date written, so user can easily understand which save is empty, which save is latest etc. Also a number of latest states can be additionally marked for visibility, e.g. with `*` or text like `latest`, `5th before latest` etc.
2. Current Save State number can be saved per-content. So, when user opens new content, current save state number will always be 0. When previously played content is loaded, save state number will be like when it was last played.
3. Something like `Current Save State is X, of Y total` can be added to notification options for when a content is loaded.

**Comment by i30817** [comment]
If this is a general savestate issue venting, i'd like to mention again that the savestate limit thing is not actually working for some weird reason.

**Comment by bslenul** [comment]
> If this is a general savestate issue venting, i'd like to mention again that the savestate limit thing is not actually working for some weird reason. Set autoincrement, set a savestate limit number, go into a game and just push on the button right past the limit.

The option is the number of saves **to keep**, it won't limit the number of slots available. So for example if you have "3", creating a state will create a .state1 file, then next time it will create a .state2 file, then .state3 and then, it will delete .state1 and create .state4, then delete .state2 and create .state5, etc.

**Comment by i30817** [comment]
Well that simply means it's impossible to tell where the savestate is when you load again if you don't have 'load last savestate' enabled.

**Comment by liandris** [comment]
Another possibility:
- Allow adding optional 1-line text notes for each state, so they are more descriptive on selection screen. Display this info on Load state notification too.

**Comment by vitran96** [comment]
(2) I would like this feature too.
With auto-increment, it is impossible to remember which save state was last saved. There is also (I think?) no way to sort the save states by saved date.

**Comment by i30817** [comment]
I was confused since I didn't ever had trouble loading the last savestate, but it may be because of this fix : https://github.com/libretro/RetroArch/pull/17391

Aka check if auto index is enabled

btw, there is a implemention of rollover now for when the option to limit the number of savestates is set now, no longer there will be savestate number 256-266 if you set yours to 10 savestate limit and spam.

**Comment by Raxior** [comment]
> 1. Save States screen can be made more useful. Specifically, each save state can have a save date written, so user can easily understand which save is empty, which save is latest etc. Also a number of latest states can be additionally marked for visibility, e.g. with `*` or text like `latest`, `5th before latest` etc.
> 2. Current Save State number can be saved per-content. So, when user opens new content, current save state number will always be 0. When previously played content is loaded, save state number will be like when it was last played.
> 3. Something like `Current Save State is X, of Y total` can be added to notification options for when a content is loaded.

i was surprised this wasn't already a feature, seconded

**Comment by alex-robbins** [comment]
When the setting to auto-increment the state slot is enabled, retroarch is supposed to detect the highest-numbered valid slot and select that when content is loaded, but this is broken in version 1.20 (the current version). That issue has been fixed on master by #17351. 

I'm not sure how it used to work when auto-increment of the state slot is disabled, but looking at #17391, for future versions it sounds like slot 0 will be selected.

**Comment by automatedbugreportingfacility** [comment]
Yes, the current state slot management is terrible and makes save states virtually unusable. How can one even tell if a slot was used or not, it's not written anywhere when a state slot number is selected AFAICT? A user just selects a number blindly and must remember whether each of 1000 slots was used and when -- is there even a human being who can remember all of that?

**Comment by i30817** [comment]
It's a defaults problem now, you can now expect the saves to cycle back so if you limit the max number, you'll never have gaps at the cost of overwriting  the oldest ones.

I'm a bit annoyed there is no way to do a dolphin style keyboard savestate method but it's not like you wouldn't need the current one anyway,  not all platforms have keyboards and controllers don't have enough buttons anyway (except maybe the overlay could have something like that, or even a "partial overlay" idea where you keep save\load state buttons but everything else is managed by the physical controller.
